### DISCUSSION TITLE
Create 3 with Arduino/ESP32

### DISCUSSION THREAD

**Discussion by vivianweli** [iRobotEducation/create3_docs#315] [Questions] [answered]
Hi everyone,
We were thinking of controlling the robot over Bluetooth with python, and interfacing the robot with ESP32, which will send some information to the robot that the robot could use. However we realized if we connect ESP32 through USB-C to the robot, we cannot control the robot using Python SDK on our laptop over Bluetooth due to the toggle switch. We have no experience in the other ways (MicroPython or ROS2) so we have some questions about how we could get around that…

1. Can we control Create3 using ESP32 in Arduino by sending opcodes through serial communication? (like Create2 open source platform)
2. Can we control Create3 using ESP32 at all? Running Python SDK through MicroPython on it (what is the full setup process)? Or running ROS2? Should we?
3. If we control Create3 using ROS2 on our laptop over wifi, can we still have a USB-C connection with ESP32 and be able to access information from it? If so, how?

Any answers/advice/corrections to our understanding will be greatly appreciated! Thanks in advance.

**Comment by gruedisueli** [comment]
I also am curious about this. In the past I worked with the Create2 and used Arduinos to connect via the DIN connection, but from what I can see, this is a bit different now. 

I did see in the docs reference to a "beta" feature for serial communication over USB.

In simple terms I just want to be able to use the same methods to communicate with a Create3 that was possible with the Create2, or something analogous to this. 

This is the only reference I can find to talking to external devices via Serial
[https://iroboteducation.github.io/create3_docs/webserver/serial-config/](https://iroboteducation.github.io/create3_docs/webserver/serial-config/)

**Comment by shamlian** [comment, marked as answer]
Right now, there is no simple UART serial-style control interface for the Create 3 robot. Your options are to use the [BLE protocol](https://github.com/iRobotEducation/root-robot-ble-protoco) using the [SDK](https://github.com/iRobotEducation/irobot-edu-python-sdk), which is likely the most micro-friendly option, or to use ROS 2 over either USB-over-Ethernet or Wi-Fi.

The serial forwarder is just that -- allowing a device connected to the Create 3 robot's USB port to be forwarded over a socket on the Create 3 robot. The use case here would be to interact with a device that only uses UART serial (Arduino, LEGO Spike Prime Hub, etc) from a remote machine. [This lesson](https://edu.irobot.com/learning-library/intro-to-ros-2-serial-communication-with-create-3) in our learning library can take you through using that, if interested.

Certainly, we're open to a feature request for other interfaces emulated over the USB-C port -- feel free to put them in as [issues](https://github.com/iRobotEducation/create3_docs/issues).

I'm trying to chew through the use case that you @vivianweli are going for -- unfortunately, you're right. If you need the serial forwarding for the ESP32, you can't also have the BLE radio, because that single USB port is multiplexed either to the BLE radio or to the USB-C port. One thought is that you could plug a USB hub into the robot's USB-C port. You could then plug a BLE radio dongle and your ESP32 as downstream devices into the hub, but the choice of dongle would be critical to ensure success. You might try the Laird BT851 if you go this route.
